# LENDING CLUD CASE STUDY

_" This  __"BANKING and FINANCIAL SERVICES"__  case study __"aims to benefit"__ consumer finance __"company to identify risky loan applicants"__. Additionally, it also __"helps to recognize the patterns which indicate if a loan applicant is likely to default"__, which could be used by the company to make decision on loan approval."_

#### Dataset 
loan.csv

#### Data Dictionary
Data_Dictionary.xlsx

## IMPORT ALL REQUIRED LIBRARIES
This step involves with importing all the required libraries for Lending Club Case Study.

In [132]:
#   Import all the required Python libraries

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings('ignore')

## SOURCING
This step involves fetching the input data/dataset for the case.

In [133]:
#   Fetch the data from loan.csv file.
loanData = pd.read_csv('loan.csv')
loanData.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

  int_rate  installment grade sub_grade  ... num_tl_90g_dpd_24m  \
0   10.65%       162.87     B        B2  ...                NaN   
1   15.27%        59.83     C        C4  ...                NaN   
2   15.96%        84.33     C        C5  ...                NaN   
3   13.49%       339.31     C        C1  ...                NaN   
4   12.69%        67.79     B        B5  ...                NaN   

  num_tl_op_past_12m pct_tl_nvr_dlq  percent_bc_gt_75 pub_rec_bankruptcies  \
0                NaN            NaN               NaN                  0.0   
1                NaN            NaN               NaN                  0.0   
2                NaN            NaN               NaN                  0.0   
3                NaN            NaN               NaN                  0.0   
4                NaN            NaN               NaN                  0.0   

  tax_liens tot_hi_cred_lim total_bal_ex_mort total_bc_limit  \
0       0.0             NaN               NaN            NaN   
1       0.0             NaN               NaN            NaN   
2       0.0             NaN               NaN            NaN   
3       0.0             NaN               NaN            NaN   
4       0.0             NaN               NaN            NaN   

  total_il_high_credit_limit  
0                        NaN  
1                        NaN  
2                        NaN  
3                        NaN  
4                        NaN  

[5 rows x 111 columns]

In [134]:
#   Print the file's head() data here.
print(loanData.head())
#   Print the file's tail() data here.
print(loanData.head())

        id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

  int_rate  installment grade sub_grade  ... num_tl_90g_dpd_24m  \
0   10.65%       162.87     B        B2  ...                NaN   
1   15.27%        59.83     C        C4  ...                NaN   
2   15.96%        84.33     C        C5  ...                NaN   
3   13.49%       339.31     C        C1  ...                NaN   
4   12.69%        67.79     B        B5  ...                NaN   

  num_tl_op_past_12m pct_tl_nvr_dlq  percent_bc_gt_75 pub_rec_bankruptcies  \
0                NaN            NaN               

In [135]:
# Lets get further information about the dataset from loan.csv.
loanData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Columns: 111 entries, id to total_il_high_credit_limit
dtypes: float64(74), int64(13), object(24)
memory usage: 33.6+ MB


###   SOURCING Analysis    
1. Shape of the dataset --> 39717 rows and 111 columns<br><br>
2. Data types scattered in the loan.csv file --> float64(74), int64(13), object(24)<br>
-There are __74__ columns with ___Float___ as data type<br>
-There are __13__ columns with ___Integers___ as as datatype<br>
-There are __24__ columns with ___Object___ as as datatype<br>

##   CLEANING
This step involves cleaning the input data/dataset for the case, which could create issue for our analysis.

In [136]:
#   Lets remove the columns with null values 
loanData = loanData.dropna(axis=1, how='all')
print(loanData.head())
print(loanData.tail())

        id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

  int_rate  installment grade sub_grade  ... next_pymnt_d last_credit_pull_d  \
0   10.65%       162.87     B        B2  ...          NaN             May-16   
1   15.27%        59.83     C        C4  ...          NaN             Sep-13   
2   15.96%        84.33     C        C5  ...          NaN             May-16   
3   13.49%       339.31     C        C1  ...          NaN             Apr-16   
4   12.69%        67.79     B        B5  ...       Jun-16             May-16   

  collections_12_mths_ex_med  policy_code applicat

In [137]:
#   Check the shape of the information of newly saved dataframe
loanData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Data columns (total 57 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   id                          39717 non-null  int64  
 1   member_id                   39717 non-null  int64  
 2   loan_amnt                   39717 non-null  int64  
 3   funded_amnt                 39717 non-null  int64  
 4   funded_amnt_inv             39717 non-null  float64
 5   term                        39717 non-null  object 
 6   int_rate                    39717 non-null  object 
 7   installment                 39717 non-null  float64
 8   grade                       39717 non-null  object 
 9   sub_grade                   39717 non-null  object 
 10  emp_title                   37258 non-null  object 
 11  emp_length                  38642 non-null  object 
 12  home_ownership              39717 non-null  object 
 13  annual_inc                  397

In [138]:
#   Remove all the duplicate items if there are any based on 'id' column 

#   Get the list of ids where only 1 id is taken if multiple found. 
loanData.duplicated(subset=None, keep='first').count()
#   Print the duplicate columns 
loanData[loanData.duplicated(['id'])]

Empty DataFrame
Columns: [id, member_id, loan_amnt, funded_amnt, funded_amnt_inv, term, int_rate, installment, grade, sub_grade, emp_title, emp_length, home_ownership, annual_inc, verification_status, issue_d, loan_status, pymnt_plan, url, desc, purpose, title, zip_code, addr_state, dti, delinq_2yrs, earliest_cr_line, inq_last_6mths, mths_since_last_delinq, mths_since_last_record, open_acc, pub_rec, revol_bal, revol_util, total_acc, initial_list_status, out_prncp, out_prncp_inv, total_pymnt, total_pymnt_inv, total_rec_prncp, total_rec_int, total_rec_late_fee, recoveries, collection_recovery_fee, last_pymnt_d, last_pymnt_amnt, next_pymnt_d, last_credit_pull_d, collections_12_mths_ex_med, policy_code, application_type, acc_now_delinq, chargeoff_within_12_mths, delinq_amnt, pub_rec_bankruptcies, tax_liens]
Index: []

[0 rows x 57 columns]

In [139]:
#   Lets Fetch further columns with only single values in all the rows. If found, removing them as these wont be able to help in our analysis

#   fetch all column names
col_list = np.array(loanData.columns.tolist())
col_list

#   Print the shape of dataframe before droping the columns with 1 value
print("Before running the logic to check for columns with only single value, dataframe shape is - "+str(loanData.shape))
#   code to loop all the columns, if the count of unique values in a column is 1, remove the column
for i in col_list:
    if len(loanData[str(i)].unique()) == 1:
        loanData = loanData.drop([str(i)], axis=1)

#   Print the shape of dataframe before droping the columns with 1 value
print("After running the logic to check for columns with only single value, dataframe shape is - "+str(loanData.shape))

print(loanData.head())
print(loanData.tail())

Before running the logic to check for columns with only single value, dataframe shape is - (39717, 57)
After running the logic to check for columns with only single value, dataframe shape is - (39717, 51)
        id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

  int_rate  installment grade sub_grade  ... recoveries  \
0   10.65%       162.87     B        B2  ...       0.00   
1   15.27%        59.83     C        C4  ...     117.08   
2   15.96%        84.33     C        C5  ...       0.00   
3   13.49%       339.31     C        C1  ...       0.00   
4   12.69%        67.79     B  

In [140]:
#   Lets Fetch further null details about all the columns of the dataset loan.csv. Furthermore, drop columns where null value percentage in a column is more than 20% 
loanData.isnull().sum()

#   Print the shape of dataframe before droping the columns with null value percentage > 20%
print("Before running the logic to check for columns with null value, dataframe shape is - "+str(loanData.shape))

#   Store column names with null value in variable
NullValCols = np.array(loanData.columns[loanData.isnull().any()].tolist())
#print("Total Column count with null values - "+str(len(NullValCols)) )

for i in NullValCols:
    #print("Column Name - "+str(i)+", Count of null values in this column - "+str(loanData[str(i)].isnull().sum()))
    #print("Column Name - "+str(i)+", Percentage of null values in this column - "+str(round((loanData[str(i)].isnull().sum()/len(loanData))*100,2)))
    #remove the columns with null value percentage > 20
    if round((loanData[str(i)].isnull().sum()/len(loanData))*100,2)>20:
        loanData = loanData.drop([str(i)], axis=1)

#   Print the shape of dataframe after droping the columns with null value percentage > 20%
print("After running the logic to check for columns with null value, dataframe shape is - "+str(loanData.shape))

print(loanData.head())
print(loanData.tail())

Before running the logic to check for columns with null value, dataframe shape is - (39717, 51)
After running the logic to check for columns with null value, dataframe shape is - (39717, 47)
        id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

  int_rate  installment grade sub_grade  ... total_rec_late_fee recoveries  \
0   10.65%       162.87     B        B2  ...               0.00       0.00   
1   15.27%        59.83     C        C4  ...               0.00     117.08   
2   15.96%        84.33     C        C5  ...               0.00       0.00   
3   13.49%       339.31     

In [141]:
#fetching the data for these columns 
col_list = np.array(loanData.columns.tolist())
print(col_list)

['id' 'member_id' 'loan_amnt' 'funded_amnt' 'funded_amnt_inv' 'term'
 'int_rate' 'installment' 'grade' 'sub_grade' 'emp_title' 'emp_length'
 'home_ownership' 'annual_inc' 'verification_status' 'issue_d'
 'loan_status' 'url' 'purpose' 'title' 'zip_code' 'addr_state' 'dti'
 'delinq_2yrs' 'earliest_cr_line' 'inq_last_6mths' 'open_acc' 'pub_rec'
 'revol_bal' 'revol_util' 'total_acc' 'out_prncp' 'out_prncp_inv'
 'total_pymnt' 'total_pymnt_inv' 'total_rec_prncp' 'total_rec_int'
 'total_rec_late_fee' 'recoveries' 'collection_recovery_fee'
 'last_pymnt_d' 'last_pymnt_amnt' 'last_credit_pull_d'
 'collections_12_mths_ex_med' 'chargeoff_within_12_mths'
 'pub_rec_bankruptcies' 'tax_liens']


In [142]:
#   Check if all id's and  and membership id's are unique value, if so, we can remove them as they won't help analysis

#   Print the shape of dataframe before checking the id/membership_id column
print("Before running the logic to check id/membership_id columns with all unique value, dataframe shape is - "+str(loanData.shape))

if len(loanData)==len(loanData['id'].unique()):
    loanData = loanData.drop(['id'], axis=1)

if len(loanData)==len(loanData['member_id'].unique()):
    loanData = loanData.drop(['member_id'], axis=1)

#   Print the shape of dataframe after checking the id/membership_id column
print("After running the logic to check id/membership_id columns with all unique value, dataframe shape is - "+str(loanData.shape))

print(loanData.head())
print(loanData.tail())

Before running the logic to check id/membership_id columns with all unique value, dataframe shape is - (39717, 47)
After running the logic to check id/membership_id columns with all unique value, dataframe shape is - (39717, 45)
   loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       5000         5000           4975.0   36 months   10.65%       162.87   
1       2500         2500           2500.0   60 months   15.27%        59.83   
2       2400         2400           2400.0   36 months   15.96%        84.33   
3      10000        10000          10000.0   36 months   13.49%       339.31   
4       3000         3000           3000.0   60 months   12.69%        67.79   

  grade sub_grade                 emp_title emp_length  ...  \
0     B        B2                       NaN  10+ years  ...   
1     C        C4                     Ryder   < 1 year  ...   
2     C        C5                       NaN  10+ years  ...   
3     C        C1       AIR RESOURCES 

In [143]:
#   Dropping list of columns which are not required/ helpful for our analysis

#   Print the shape of dataframe before removing the columns which are not required
print("Before removing the columns which are not required, dataframe shape is - "+str(loanData.shape))

loanData = loanData.drop(['funded_amnt_inv', 'url', 'emp_title', 'title', 'pub_rec', 'chargeoff_within_12_mths', 'tax_liens', 'collections_12_mths_ex_med', 'out_prncp_inv', 'total_pymnt_inv', 'earliest_cr_line', 'out_prncp', 'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee', 'collection_recovery_fee', 'last_pymnt_d', 'last_pymnt_amnt', 'last_credit_pull_d'],axis=1)

#   Print the shape of dataframe after removing the columns which are not required
print("After removing the columns which are not required, dataframe shape is - "+str(loanData.shape))

print(loanData.head())
print(loanData.tail())

Before removing the columns which are not required, dataframe shape is - (39717, 45)
After removing the columns which are not required, dataframe shape is - (39717, 26)
   loan_amnt  funded_amnt        term int_rate  installment grade sub_grade  \
0       5000         5000   36 months   10.65%       162.87     B        B2   
1       2500         2500   60 months   15.27%        59.83     C        C4   
2       2400         2400   36 months   15.96%        84.33     C        C5   
3      10000        10000   36 months   13.49%       339.31     C        C1   
4       3000         3000   60 months   12.69%        67.79     B        B5   

  emp_length home_ownership  annual_inc  ...    dti delinq_2yrs  \
0  10+ years           RENT     24000.0  ...  27.65           0   
1   < 1 year           RENT     30000.0  ...   1.00           0   
2  10+ years           RENT     12252.0  ...   8.72           0   
3  10+ years           RENT     49200.0  ...  20.00           0   
4     1 year         

In [144]:
# Further remove the Null values
loanData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Data columns (total 26 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   loan_amnt             39717 non-null  int64  
 1   funded_amnt           39717 non-null  int64  
 2   term                  39717 non-null  object 
 3   int_rate              39717 non-null  object 
 4   installment           39717 non-null  float64
 5   grade                 39717 non-null  object 
 6   sub_grade             39717 non-null  object 
 7   emp_length            38642 non-null  object 
 8   home_ownership        39717 non-null  object 
 9   annual_inc            39717 non-null  float64
 10  verification_status   39717 non-null  object 
 11  issue_d               39717 non-null  object 
 12  loan_status           39717 non-null  object 
 13  purpose               39717 non-null  object 
 14  zip_code              39717 non-null  object 
 15  addr_state         

In [145]:
# #Standardizing data

# loanData.int_rate = pd.to_numeric(loanData.int_rate.apply(lambda x : x.split('%')[0]))
# loanData.emp_length = pd.to_numeric(loanData.emp_length.apply(lambda x: 0 if "<" in str(x) else (str(x).split('+')[0] if "+" in str(x) else str(x).split()[0])))
# loanData.term = pd.to_numeric(loanData.term.apply(lambda x : str(x).replace("months","").strip()))
# loanData.zip_code = pd.to_numeric(loanData.zip_code.apply(lambda x : str(x).replace("x","").strip()))

# loanData.head()
# loanData.tail()

### Cleaning Analysis

dtypes: float64(20), int64(13), object(24)<br>
No duplicate rows found on basis of id column<br>
There were 6 columns which were dropped as those had only 1 value for all rows, which was not helpful for the analysis<br>
We can see for these 4 columns, named as  'mths_since_last_record', 'mths_since_last_delinq', 'desc' and 'next_pymnt_d', has the null value percentage greater than 20%. Therefore, we have to drop these above columns.<br> 
Dropped 'id' and 'membership_id' as they have unique values across the table and couldn't help us further with the analysis<br>
After deleting the non required columns, we are left with 26 columns